### PR TITLE
Houlahan/feature/media

### PR DESCRIPTION
**Pull request by randyhoulahan**

### CODE REVIEW

File: app/views/kronos/participation-form.js [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+        getOrg()
+      })
+    }
+
+    function initStepsContacts(){
+      return getOrg().then(function(){
+        var headP  = getHead()
+        var focalP = getFocalPoint()
+        return Promise.all([headP,focalP]).then(function(){
+          $timeout(function(){
```
Comment (blaisefonseca) [inline app/views/kronos/participation-form.js]: not sure why timeout is needed?

File: app/views/kronos/participation-form.js [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+                      delete(res.data[0].meta)
+                      return _ctrl.doc = res.data[0]
+                    }
+                    return null
+                 }).catch(function(e){
+                   _ctrl.error=e
+                 })
+    }
+
+    $scope.$applyAsync(function(){
```
Comment (blaisefonseca) [inline app/views/kronos/participation-form.js]: i will use timeout over applyasync

File: app/directives/file.js [hunk 3]
```diff
@@ -1,66 +1,144 @@
-define(['app', 'text!./file.html'], function(app, fileDropTemplate) { 'use strict';
+define(['app', 'angular', 'text!./file.html'], function(app, ng, fileDropTemplate) { 'use strict';
 
-	app.directive('type', [function() {
+	app.directive('type', ['$http', function($http) {
 	    return {
 	        restrict: 'A',
 	        replace: true,
-	        require: '?ngModel', 
+            require: '?ngModel',
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/directives/file.js]: @sbilodeau ?

File: app/directives/file.js [hunk 4]
```diff
@@ -1,66 +1,144 @@
...
+                        for(var i=0; i<htmlFiles.length; ++i) {
+
+                            var formData = new FormData();
+                            var htmlFile = htmlFiles[i];
+
+                            formData.append("file", htmlFile);
+
+                            var qs = {};
+
+                            if(attr.encrypt!==undefined)
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/directives/file.js]: @sbilodeau ?

File: app/directives/kronos/address.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+			restrict : "E",
+			template : html,
+      replace: true,
+      scope: {
+        type:'=',
+        binding: "=ngModel",
+      },
+			link: function ($scope) {
+        $http.get('/api/v2015/countries',{ cache: true })
+          .then(function(o){return $filter('orderBy')(o.data, 'name.en');})
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/directives/kronos/address.js]: fixed

File: app/views/kronos/participation-form.js [hunk 6]
```diff
@@ -0,0 +1,536 @@
...
+    _ctrl.genMeta=genMeta
+    function genMeta(){
+      return {requestid:_ctrl.doc._id,conferencecode:_ctrl.conferenceCode}
+    }
+    _ctrl.viewAttachment=viewAttachment
+    function viewAttachment(url){
+      if(!url)return''
+      if(~url.indexOf('cbd.documents.temporary')||~url.indexOf('temporary-files'))
+        return url
+      return '/api/v2018/kronos/participation/requests/attachement/'+encodeURIComponent(url)
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/views/kronos/participation-form.js]: fixed

File: app/views/kronos/participation-form.js [hunk 7]
```diff
@@ -0,0 +1,536 @@
...
+
+    }
+
+
+
+    function loadPresigned(url){
+      var atts = _ctrl.organization.attachment
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/views/kronos/participation-form.js]: fixed

File: app/views/kronos/participation-form.js (hunk 1 above)
Comment (randyhoulahan) [reply app/views/kronos/participation-form.js]: fixed

File: app/views/kronos/participation-form.js [hunk 8]
```diff
@@ -0,0 +1,536 @@
...
+     $http.get('/api/v2013/thesaurus/domains/ISO639-2/terms',{ cache: true })
+      .then(function(res){
+        _ctrl.languages = res.data
+        getOrg()
+      })
+    }
+
+    function initStepsContacts(){
+      return getOrg().then(function(){
+        var headP  = getHead()
```
[the remark this answers is not in the document]
Comment (randyhoulahan) [reply app/views/kronos/participation-form.js]: fixed

File: app/views/kronos/participation-form.js (hunk 2 above)
Comment (randyhoulahan) [reply app/views/kronos/participation-form.js]: fixed

File: app/directives/file.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/directives/file.js]: yes. is only the presence on the attribute on the element. Like `disabled` on input. if you prefer I can switch to ``` scope: { encrypt: "<" }``` instead

File: app/directives/file.js [hunk 9]
```diff
@@ -1,66 +1,144 @@
...
+
+                                err = err.data || err;
+
+                                files = files.concat([{ error: err }]);
+
+                                setViewValue(files);
+
+                                onUpload(htmlFile, null, err);
+
+                                res.data;
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/directives/file.js]: no. mistake

File: app/directives/file.js [hunk 10]
```diff
@@ -1,66 +1,144 @@
...
                     return element.attr('multiple')!==undefined; 
                 }
+                
+                function isAutoUpload() { 
+                    return attr.onUpload!==undefined; 
+                }  
+                
+                function onUpload(htmlFile, file, err) { 
+
+                    return $scope.onUpload({
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/directives/file.js]: No. Use of when using `&` on directive scope  systematically generate a function even of the attribute is not set on the directive element.  
```scope: {
   onUpload : "&onUpload"
}```

File: app/directives/kronos/participant.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+          if(!$scope.isContact) $scope.attending  = {val:true}
+          $scope.binding.useOrganizationAddress = true
+          $scope.binding.requestId = $scope.requestId
+          $scope.binding.requestType = $scope.typee
+          $scope.binding.organization = $scope.organization._id
+        }
+
+        function isMedia(media){
+          for(var i=0; i<$scope.organization.medium.length; i++)
+            if($scope.organization.medium[0]===media)
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/directives/kronos/participant.js]: `return $scope.organization.medium[i]===media` cannot work with this logic. you need to test with `i` until 1st entry `$scope.organization.medium[i]===media` before returning `true`. unless `$scope.organization.medium.length` is always 1

File: app/directives/kronos/participant.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+
+        $http.get('/api/v2013/thesaurus/domains/ISO639-2/terms',{ cache: true })
+            .then(function(res){$scope.languages = res.data})
+
+        var params = {
+                        s:{'name.en':1},
+                        f:{'name.en':1,code:1}
+                      }
+
+        $http.get('/api/v2015/countries',{ params : params })
```
Comment (sbilodeau) [inline app/directives/kronos/participant.js]: This is not a comment about this commit but why do we have `/api/v2015/countries` and `/api/v2013/countries`?

File: app/routes.js [hunk 13]
```diff
@@ -38,6 +38,10 @@ define(['app', 'jquery', 'lodash', 'text!./redirect-dialog.html','providers/exte
         if(/^\/executive-secretary($|\/.*)/.test(locationPath))
             registerRoutes_esPages($routeProvider);
 
+        //es/*
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/routes.js]: @blaisefonseca you asking too many questions!!!

File: app/directives/kronos/user-messages.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+define(['app', 'text!./user-messages.html'], function(app, html) { 'use strict';
+
+	return app.directive('userMessage', ['$interpolate',function($interpolate) {
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/directives/kronos/user-messages.js]: this is the interpolation service that replace {{value}} with the corresponding text

File: app/services/conference-service.js [hunk 15]
```diff
@@ -36,24 +36,70 @@
...
                                     });
-                
+
+            }
+            function getFuture(){
+              var query = {
+                            EndDate:{ $gt: { $date: new Date()  } },
+                            institution:"CBD"
+                          }
+              return $http.get('/api/v2016/conferences', {params : { q : query, f:{ StartDate:1,MajorEventIDs:1,Title:1,Venua:1,code:1,Description:1}, s: { StartDate: 1}}, cache:true}).then(
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/services/conference-service.js]: @blaisefonseca we're now coding in spanish

File: app/services/conference-service.js [hunk 16]
```diff
@@ -36,24 +36,70 @@
...
+            }
+            function getConference(codeOrId){
+              var isOID = RegExp(/^[0-9a-fA-F]{24}$/)
+              var query = {}
+              if(isOID.test(codeOrId))
+                query._id = {'$oid':codeOrId}
+              else
+                query.code = codeOrId
+
+              return $http.get('/api/v2016/conferences', {params : { q : query, f:{ StartDate:1,MajorEventIDs:1,Title:1,Venua:1,code:1,Description:1}, s: { StartDate: 1}}, cache:true}).then(
```
[the remark this answers is not in the document]
Comment (sbilodeau) [reply app/services/conference-service.js]: @blaisefonseca SPANISH

File: app/views/kronos/participation-form-finished.html [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+      <br>
+      <br>
+      <h1>Finished</h1>
+    </div>
+    <div class="col-xs-10 col-xs-offset-1">
+
+      <p>Thank you for your media accreditation submission.
+        Please note that your request will be processed within 7 days under normal circumstances.
+         If you do not receive a response within this timeframe, please contact Johan Hedlund at
+         <a href="mailto:johan.hedlund@cbd.int">johan.hedlund@cbd.int</a>.
```
Comment (sbilodeau) [inline app/views/kronos/participation-form-finished.html]: I suggest that we create an alias email for this: may be `media@cbd.int` (i'm inspired)

File: app/views/kronos/participation-form-org-header.html [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+      <h4>Contacts</h4>
+      <hr/>
+    </div>
+    <div class="col-xs-2">
+      Head
+    </div>
+    <div class="col-xs-9">
+        Randy J. Houlahan<br>
+        Executive Secretary,<br>
+        <a href="mailto:dddd@sss.com">randy.houlahan@cbd.int</a>
```
Comment (sbilodeau) [inline app/views/kronos/participation-form-org-header.html]: this is wrong `dddd@sss.com`

File: app/views/kronos/participation-form-org-header.html [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    <div class="col-xs-12">
+      <hr/>
+    </div>
+    <div class="col-xs-2">
+      Focal Point
+    </div>
+    <div class="col-xs-9">
+        Blaise Fonseca<br>
+        Lead Tabber,<br>
+        <a href="mailto:dddd@sss.com">Blaise.Fonseca@cbd.int</a>
```
Comment (sbilodeau) [inline app/views/kronos/participation-form-org-header.html]: this is worst `Blaise.Fonseca@cbd.int`

File: app/directives/kronos/user-messages.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (blaisefonseca) [reply app/directives/kronos/user-messages.js]: i thought angularjs does the interpolation for us

File: app/views/kronos/participation-form-finished.html (hunk 17 above)
Comment (randyhoulahan) [reply app/views/kronos/participation-form-finished.html]: check

File: app/directives/kronos/participant.js [hunk 20]
```diff
@@ -75,12 +81,15 @@ define(['app', 'text!./participant.html','./address','services/conference-servic
         }
 
         function save(){
+          if($scope.binding && $scope.binding.meeting && !$scope.binding.meeting.length)
+            delete($scope.binding.meeting)
```
Comment (blaisefonseca) [inline app/directives/kronos/participant.js]: delete?

**Pull request merged**